README.md
# port-utils

## openFreePort

This is another version of `warp`'s `openFreePort` function. Nice if you don't already depend on `warp`.

`openFreePort` returns a socket on random port and the port it has been bound to.

```haskell
openFreePort :: IO (Int, Socket)
```

## wait

`wait` will attempt to connect to a host and port until successful. Between each unsuccessful attempt it sleeps for 10 ms

Here is an example of the primary function:

```haskell
import Network.Socket.Wait (wait)

void $ forkIO $ Warp.run 7000 app
-- Wait for the server to start listening on the socket
wait "127.0.0.1" 7000
-- Communicate with the server
...
```

In bash one could write:

```bash
while ! nc -z localhost 7000 ; do sleep 0.01 ; done
```

The above was copied from this stackoverflow answer https://stackoverflow.com/a/50008755

ChangeLog.md
# Changelog for wait-on-port
- 0.0.0.1 First version


test/Tests/Network/Socket/WaitSpec.hs
{-# OPTIONS_GHC -fno-warn-orphans #-}
module Tests.Network.Socket.WaitSpec where
import qualified Network.Socket.Wait.Internal as W
import qualified Network.Socket.Free as F
import qualified Network.Socket as N
import qualified Control.Concurrent as C
import qualified Control.Concurrent.STM as S
import qualified Control.Exception as E
import qualified Test.Hspec as H
import qualified System.Timeout as T
import qualified Control.Concurrent.Async as A
import qualified Control.Monad.Trans.State as St

instance Monad m => Semigroup (St.StateT s m ()) where
  (<>) = (>>)

instance Monad m => Monoid (St.StateT s m ()) where
  mempty = pure ()
  mappend = (<>)

main :: IO ()
main = H.hspec spec

spec :: H.Spec
spec = do
  H.describe "wait" $ H.before F.openFreePort $ H.after (N.close . snd) $ do
    H.it "does not connect if port is unavailable" $ \_ -> do

      T.timeout 100000 (W.wait "127.0.0.1" 0) >>= \case
        Nothing -> pure ()
        Just () -> fail "wait returned! I should have blocked forever!"
    H.it "does connect if port is available" $ \(port, sock) -> do
      N.listen sock 128

      T.timeout 100000 (W.wait "127.0.0.1" port) >>= \case
        Nothing -> fail "wait timed out attempting to connect! I should have returned immediantly!"
        Just () -> pure ()

  H.describe "connectAction" $ H.before F.openFreePort $ H.after (N.close . snd) $ do
    H.it "returns False if it fails to connect" $ \_ ->
      W.connectAction "127.0.0.1" 0 `H.shouldReturn` False

    H.it "returns True if it connects" $ \(port, sock) -> do
      N.listen sock 128
      W.connectAction "127.0.0.1" port `H.shouldReturn` True

    H.it "returns True if connects after failing" $ \(port, sock) -> do
      N.close sock
      -- small race here if another process opens on this port

      W.connectAction "127.0.0.1" port `H.shouldReturn` False

      E.bracket (N.socket N.AF_INET N.Stream N.defaultProtocol) N.close $ \sock' -> do
         N.bind sock' $ N.SockAddrInet (fromIntegral port) $ N.tupleToHostAddress (127,0,0,1)
         N.listen sock' 128

         W.connectAction "127.0.0.1" port `H.shouldReturn` True

    H.it "throws if the host does not exist" $ \(_, _) -> do
      W.connectAction "invalid." 3000 `H.shouldThrow` (\(_ :: IOError) -> True)

  H.describe "waitM" $ do
    H.it "returns immediantly if the action returns True" $
      flip St.execState False (W.waitM mempty (St.put True) (pure True)) `H.shouldBe` False

    H.it "loops until the action returns True" $ do
      let theAction = do
            St.get >>= \case
              0 -> St.put 1 >> pure False
              _ -> St.put 2 >> pure True

      flip St.execState 0 (W.waitM mempty (pure ()) theAction) `H.shouldBe` 2


src/Network/Socket/Free.hs
module Network.Socket.Free (openFreePort) where
import qualified Network.Socket as N
import qualified Control.Concurrent as C
import qualified Control.Exception as E
import qualified System.IO.Error as Error

-- | Open a TCP socket on a random free port. This is like 'warp''s
--   openFreePort.
--
--   Since 0.0.0.1
openFreePort :: IO (Int, N.Socket)
openFreePort =
  E.bracketOnError (N.socket N.AF_INET N.Stream N.defaultProtocol) N.close
    $ \sock -> do
      N.bind sock $ N.SockAddrInet 0 $ N.tupleToHostAddress (127,0,0,1)
      N.getSocketName sock >>= \case
        N.SockAddrInet port _ -> pure (fromIntegral port, sock)
        addr -> E.throwIO
          $ Error.mkIOError Error.userErrorType
            (  "openFreePort was unable to create socket with a SockAddrInet. "
            <> "Got " <> show addr
            )
            Nothing
            Nothing


src/Network/Socket/Wait.hs
module Network.Socket.Wait
  ( -- * Simple Wait Api
    wait
    -- * Advanced Wait Api
  , waitWith
  , EventHandlers (..)
  , defaultDelay
  ) where
import qualified Network.Socket as N
import qualified Control.Concurrent as C
import qualified Control.Exception as E

-------------------------------------------------------------------------------
-- Simple Api
-------------------------------------------------------------------------------

-- | 'wait' will attempt to connect to the given host and port repeated every
-- 10 milliseconds until it is successful. It will throw an 'IOError' if the
-- host cannot be resolved.
--
-- A typical use case is to call 'wait' in test code to wait for a server to
-- start before trying to connect. For example:
--
-- @
--    void $ forkIO $ Warp.run 7000 app
--    -- Wait for the server to start listening on the socket
--    wait "127.0.0.1" 7000
--    -- Communicate with the server
-- @
--
-- If you would like to control the delay or understand how many connection
-- attempts were made use 'waitWith'.
--
-- Since 0.0.0.1
wait :: String
     -- ^ Host
     -> Int
     -- ^ Port
     -> IO ()
wait = waitWith mempty defaultDelay

-------------------------------------------------------------------------------
-- Advanced Api
-------------------------------------------------------------------------------
-- | The default delay between retries is 10000 microseconds (10 ms)
defaultDelay :: Int
defaultDelay = 10000

-- | The 'EventHandlers' is a record of 'IO' actions that are called when
--   interesting events occur in the lifecycle of the 'waitWith' loop.
--   One can pass in custom 'EventHandlers' values to implement logging
--   and other forms of instrumentation.
--
--   For example for debug one could print out each step using:
--
-- @
--    printHandlers = EventHandlers
--      { acting     = putStrLn "acting"
--      , delaying   = putStrLn "delaying"
--      , restarting = putStrLn "restarting"
--      }
-- @
--
data EventHandlers m = EventHandlers
  { acting :: m ()
  -- ^ Called after the socket is created
  , delaying :: m ()
  -- ^ Called after a failed attempt to connect before the thread
  -- is put to sleep.
  , restarting :: m ()
  -- ^ Called before a recursive call to restart the connection attempt
  }

instance Semigroup (m ()) => Semigroup (EventHandlers m) where
  x <> y = EventHandlers
    { acting        = acting      x <> acting     y
    , delaying      = delaying    x <> delaying   y
    , restarting    = restarting  x <> restarting y
    }

instance Monoid (m ()) => Monoid (EventHandlers m) where
  mempty  = EventHandlers mempty mempty mempty
  mappend = (<>)

-- | Advanced usage. In most situations calling 'wait' will suffice. This allows
-- one to customize the delay between retries and debug the behavior of the
-- function. 'wait' is defined as
--
-- @
--    wait = waitWith mempty defaultDelay
-- @
--
-- Since 0.0.0.1
waitWith :: EventHandlers IO
         -- ^ A record of IO actions that are called during sp
         -> Int
         -- ^ Microseconds to delay
         -> String
         -- ^ Host
         -> Int
         -- ^ Port
         -> IO ()
waitWith eh delay host port
  = waitM eh (C.threadDelay delay) (connectAction host port)

waitM :: Monad m => EventHandlers m -> m () -> m Bool -> m ()
waitM eh@EventHandlers {..} delayer action = do
  acting
  action >>= \case
    True -> pure ()
    False -> do
      delaying
      delayer

      restarting
      waitM eh delayer action

connectAction :: String -> Int -> IO Bool
connectAction host port = do
  let hints = N.defaultHints { N.addrSocketType = N.Stream }
  -- getAddrInfo returns a non-empty array or throws per the doc
  addr:_ <- N.getAddrInfo (Just hints) (Just host) (Just $ show port)
  e <- E.try $ E.bracket
        (N.socket (N.addrFamily addr) (N.addrSocketType addr) (N.addrProtocol addr))
        N.close
        $ \sock -> N.connect sock $ N.addrAddress addr

  pure $ case e of
    Left (_ :: IOError) -> False
    Right _ -> True


src/Network/Socket/Wait/Internal.hs
{-|
  This internal modules exposes additional functions for testing.
  Use 'Network.Socket.Wait' instead
|-}
module Network.Socket.Wait.Internal
  ( -- * Simple Wait Api
    wait
    -- * Advanced Wait Api
  , waitWith
  , EventHandlers (..)
  , defaultDelay
  , connectAction
  , waitM
  ) where
import qualified Network.Socket as N
import qualified Control.Concurrent as C
import qualified Control.Exception as E

-------------------------------------------------------------------------------
-- Simple Api
-------------------------------------------------------------------------------

-- | 'wait' will attempt to connect to the given host and port repeated every
-- 10 milliseconds until it is successful. It will throw an 'IOError' if the
-- host cannot be resolved.
--
-- A typical use case is to call 'wait' in test code to wait for a server to
-- start before trying to connect. For example:
--
-- @
--    void $ forkIO $ Warp.run 7000 app
--    -- Wait for the server to start listening on the socket
--    wait "127.0.0.1" 7000
--    -- Communicate with the server
-- @
--
-- If you would like to control the delay or understand how many connection
-- attempts were made use 'waitWith'.
--
-- Since 0.0.0.1
wait :: String
     -- ^ Host
     -> Int
     -- ^ Port
     -> IO ()
wait = waitWith mempty defaultDelay

-------------------------------------------------------------------------------
-- Advanced Api
-------------------------------------------------------------------------------
-- | The default delay between retries is 10000 microseconds (10 ms)
defaultDelay :: Int
defaultDelay = 10000

-- | The 'EventHandlers' is a record of 'IO' actions that are called when
--   interesting events occur in the lifecycle of the 'waitWith' loop.
--   One can pass in custom 'EventHandlers' values to implement logging
--   and other forms of instrumentation.
--
--   For example for debug one could print out each step using:
--
-- @
--    printHandlers = EventHandlers
--      { acting     = putStrLn "acting"
--      , delaying   = putStrLn "delaying"
--      , restarting = putStrLn "restarting"
--      }
-- @
--
data EventHandlers m = EventHandlers
  { connecting :: m ()
  -- ^ Called before the socket is created for connection
  , delaying :: m ()
  -- ^ Called after a failed attempt to connect before the thread
  -- is put to sleep.
  , restarting :: m ()
  -- ^ Called before a recursive call to restart the connection attempt
  }

instance Semigroup (m ()) => Semigroup (EventHandlers m) where
  x <> y = EventHandlers
    { connecting = connecting x <> connecting y
    , delaying   = delaying   x <> delaying   y
    , restarting = restarting x <> restarting y
    }

instance Monoid (m ()) => Monoid (EventHandlers m) where
  mempty  = EventHandlers mempty mempty mempty
  mappend = (<>)

-- | Advanced usage. In most situations calling 'wait' will suffice. This allows
-- one to customize the delay between retries and debug the behavior of the
-- function. 'wait' is defined as
--
-- @
--    wait = waitWith mempty defaultDelay
-- @
--
-- Since 0.2.0.0
waitWith :: EventHandlers IO
         -- ^ A record of IO actions that are called during sp
         -> Int
         -- ^ Microseconds to delay
         -> String
         -- ^ Host
         -> Int
         -- ^ Port
         -> IO ()
waitWith eh delay host port
  = waitM eh (C.threadDelay delay) (connectAction host port)

-- | This function loops if the second argument returns 'False'. Between the
--   recursive calls it will call it's first argument.
waitM :: Monad m => EventHandlers m -> m () -> m Bool -> m ()
waitM eh@EventHandlers {..} delayer action = do
  connecting
  action >>= \case
    True -> pure ()
    False -> do
      delaying
      delayer

      restarting
      waitM eh delayer action

-- | This function attempts to initiate a TCP connection to a given host and
--   port. If the host does not exist it throws an IOError. If the connection
--   fails it returns a 'False'. If it connects it returns 'True'.
connectAction :: String -> Int -> IO Bool
connectAction host port = do
  let hints = N.defaultHints { N.addrSocketType = N.Stream }
  -- getAddrInfo returns a non-empty array or throws per the doc
  addr:_ <- N.getAddrInfo (Just hints) (Just host) (Just $ show port)
  e <- E.try $ E.bracket
        (N.socket (N.addrFamily addr) (N.addrSocketType addr) (N.addrProtocol addr))
        N.close
        $ \sock -> N.connect sock $ N.addrAddress addr

  pure $ case e of
    Left (_ :: IOError) -> False
    Right _ -> True
